ADR/adr-0008-environment-provisioning.md

```diff
@@ -0,0 +1,464 @@
+parameters section can be used to forward additional information to the provisioner. The
+reclaimPolicy field will tell the provisioner what to do with the [DT] once its corresponding [DTC] is
+deleted, the values can be Retain or Delete.
+
+**Immutable object**: yes
+
+**Scope**: cluster
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: DeploymentTargetClass
+metadata:
+    name: isolation-level-namespace
+spec:
+    provisioner: appstudio.redhat.com/devsandbox
+    parameters: {}
+    reclaimPolicy: Delete
+```
+
+#### Environment
+
+[Environment] objects refer to a [DTC] using the `deploymentTargetClaim`
+field. The environment controller will wait for the [DTC] to get to the
+`bound` phase, once it is bound, it will reach the [DT] and read the
+target's connection details from the kubernetesCredentials field and
+configure Argo/Gitops services to use them.
+
+**Immutable object**: no
+
+**Scope**: namespace
+
+```yaml
+apiVersion: appstudio.redhat.com/v1alpha1
+kind: Environment
+metadata:
+    name: prod
+spec:
+    displayName: “Production for Team A”
+    deploymentStrategy: AppStudioAutomated
+    parentEnvironment: staging
+    tags: [prod]
+    configuration:
+        target:
+            deploymentTargetClaim:
+                claimName: prod-dtc
+```
+
+### Controllers
+
+#### DeploymentTargetBinder
+
+Binds [DeploymentTargetClaim] to a [DeploymentTarget] that satisfies its requirements.
+
+It watches for [DTC] and objects and tries to find a matching [DT] for each one of them. In addition, it's responsible for updating the `Status` sections of the [DT] and [DTC] (it also watches [DTC] objects). It marks [DTC] objects that requires dynamic provisioning.
+
+A [DT] that was created dynamically for a specific [DTC] will always be attached to it.
+
+[DT] and [DTC] have one to one bindings.
+
+##### Binding Loop (DTC Reconciliation)
+
+![](../diagrams/ADR-0008/binding-controller.jpg)
+
+#### DeploymentTargetProvisioner
+
+Watch for [DTC] objects. If the [DTC] was marked for a dynamic provisioning by the provisioner, it reads the parameters from the [DTCLS] mentioned in the [DTC], provisions the target and creates a [DT]
+object which references the [DTC] that started the process.
+
+When a [DTC] is deleted, if it was bound to a [DT] created by the provisioner, it reclaims the [DT] and
+the actual cluster that was created for it based on the reclaimPolicy configuration.
+
+##### Provision Loop (DTC Reconciliation)
+
+![](../diagrams/ADR-0008/provision-loop.jpg)
+
+##### Deprovision Loop (DT Reconciliation)
+
+![](../diagrams/ADR-0008/deprovision-loop.jpg)
+
+#### EnvironmentController
+
+Watch for [Environment] objects. If the [Environment] references a [DTC] that is `Bound` and that
+[DTC] references a [DT] that contains a `spec.kubernetesCredentials.clusterCredentialsSecret` field,
+then ensure that there exists a [GitOpsDeploymentManagedEnvironment] resource that also references
```

Comment:
I think it's better to have two different API's with a controller that bridges between them rather than a single API that will create a coupling between the controllers. It's expected that in the future the DT will contain more details about the cluster/namespace it represents, for example, its CPU architecture, and I don't think that this information should interest the GitopsService.

Reply:
Thank you @gbenhaim, fair point.

The reason I feel having many APIs which wrap cluster credentials is that there is a long-term API-maintenance overhead even though short-term it is quicker to get things done. 

The description of a Kubernetes deployment target (Cluster properties and cluster credentials) would always be the same - with a subset of properties being optional.